hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -142,31 +219,58 @@ struct Visitor : hilti::visitor::PreOrder<bool, Visitor> {
     }
 
     result_t operator()(const declaration::Function& x, position_t p) {
+        const auto function_id = function_identifier(x, p);
+        const auto& [module_id, struct_id, field_id] = function_id;
+
+        if ( ! goodHook(function_id) )
+            return false;
+
         switch ( _stage ) {
             case Stage::COLLECT: {
-                auto module_id = x.id().sub(-3);
-                if ( module_id.empty() ) {
-                    // HILTI hook functions do not include the name their module in their ID.
-                    if ( auto module = p.findParent<Module>() )
-                        module_id = module->get().id();
-                }
-
-                auto struct_id = x.id().sub(-2);
-                auto field_id = x.id().sub(-1);
+                // Record this hook as declared if it is not already known.
+                auto& function = (*_functions)[function_id];
+                function.declared = true;
+
+                for ( auto&& fn : x.childsOfType<Function>() ) {
```

Comment:
There should be one function at most, and that one you get with `function()`, no?